server.js

```diff
@@ -14,10 +12,103 @@ const welcomeMessage = {
 //This array is our "data store".
 //We will start with one message in the array.
 //Note: messages will be lost when Glitch restarts our server.
+
+app.use(cors());
+app.use(express.json());
+app.use(bodyParser.json());
+
+// parse application/x-www-form-urlencoded - for accessing form
+app.use(bodyParser.urlencoded({ extended: false }));
+
 const messages = [welcomeMessage];
 
 app.get("/", function (request, response) {
   response.sendFile(__dirname + "/index.html");
+  response.status(200).send("Welcome to my message App");
 });
 
-app.listen(process.env.PORT);
+/*to get all the messages*/
+app.get("/messages", function (request, response) {
+  response.status(200).send({ messages });
+});
+
+/*to post a message*/
+app.post("/messages", function (req, res) {
+  const newMessage = req.body;
+  /*access Id */
+  const previousMessage = messages[messages.length - 1];
+  const currentId = previousMessage.id + 1;
+
+  /*merged objects in the final one*/
+  const newMessageWithId = { ...newMessage, id: currentId };
+
+  /*validate message text and from properties*/
+  if (!newMessageWithId.text || !newMessageWithId.from) {
+    return res.status(400).json({
+      error: "Both name and message are required",
+    });
+  }
+
+  messages.push(newMessageWithId);
+  res.status(201).send({ newMessageWithId });
+});
+
+/*To get by ID*/
+app.get("/messages/:id", function (request, response) {
+  // console.log(request.params, "<------- req params");
+  const idToFind = Number(request.params.id);
+  const message = messages.find((message) => message.id === idToFind);
+  console.log(message);
+  response.status(200).send({ message });
+});
+
+/*to delete a message by ID */
+app.delete("/messages/:id", function (request, response) {
+  const idToFind = Number(request.params.id);
+  const message = messages.find((message) => message.id === idToFind);
+  const indexOfMessage = messages.indexOf(message);
+
+  if (indexOfMessage !== -1) {
+    messages.splice(indexOfMessage, 1);
+    response
+      .status(200)
+      .send({ message: `Message with ID ${indexOfMessage} has been deleted` });
+  } else {
+    response
+      .status(404)
+      .send({ message: `Message with ID ${indexOfMessage} not found` });
+  }
+
+  // response.status(200).send({ message });
+});
+
+/*to search messages by text*/
+app.get("/messages/search", function (req, res) {
+  const searchText = req.query.text;
+  const matchingMessages = messages.filter(
+    (message) => message.text.indexOf(searchText) !== -1
+  );
+  res.status(200).send({ messages: matchingMessages });
+});
+
+/*to get the latest 10 messages*/
+app.get("/messages/latest", function (req, res) {
+  const latestMessages = messages.slice(-10);
+  res.status(200).send({ messages: latestMessages });
+});
+
+app.put("/messages/:id", function (req, res) {
```

Comment:
What if no message with the specified id is present? Also, what if the provided "id" is not an integer?

Reply:
Wow, didn't think about it. Added  ```
if (messageIndex === -1) {
    return res.status(404).send({ error: "Message not found" });
  }
``` and ```
if (isNaN(id)) {
    return res.status(400).send({ error: "Invalid ID" });
  }
```. Do you think it's a good way to address these issues, or is there a better way?